test/hotspot/jtreg/gc/metaspace/TestMetaspacePerfCounters.java

```diff
@@ -204,32 +239,26 @@ public static void main(String[] args) throws Exception {
     }
 
     private static void checkPerfCounters(String ns) throws Exception {
-        long gcCountBefore;
-        long gcCountAfter;
-        long minCapacity;
-        long maxCapacity;
-        long capacity;
-        long used;
-
-        // The perf counter values are updated during GC and to be able to
-        // do the assertions below we need to ensure that the values are from
-        // the same GC cycle.
-        do {
-            gcCountBefore = currentGCCount();
+        PerfCounterSnapshot snap1 = new PerfCounterSnapshot();
+        PerfCounterSnapshot snap2 = new PerfCounterSnapshot();
 
-            minCapacity = getMinCapacity(ns);
-            maxCapacity = getMaxCapacity(ns);
-            capacity = getCapacity(ns);
-            used = getUsed(ns);
+        final int MaxAttempts = 10;
 
-            gcCountAfter = currentGCCount();
-            assertGTE(gcCountAfter, gcCountBefore);
-        } while(gcCountAfter > gcCountBefore);
+        int attempts = 0;
+        do {
+            snap1.get(ns);
+            VarHandle.fullFence();
+            snap2.get(ns);
+        } while ((!snap1.equals(snap2)) && (++attempts < MaxAttempts));
+
+        if (attempts == MaxAttempts) {
+            throw new Exception("Failed to get stable reading of metaspace performance counters after " + MaxAttempts + " tries");
+        }
```

Comment:
I think we should remove the second test in the while clause, and instead move the `throw` statement (with if) into the loop. If we do that we get rid of one test, and it is more clear that `++attempts < MaxAttempts` matches `attempts == MaxAttempts`. If we do that, we can also move `MaxAttempts` and `attempts` into the loop, limiting their life time. If MaxAttempts is only used once, I also think it can be removed, but do keep it if you or Kim want it.

Reply:
Sorry, `attempts` can of course not be moved into the loop of course.

Reply:
or maybe use a for loop:
```
for {int attempts = 0; ; attempts++) {
  snap1.get(ns);
  VarHandle.fullFence();
  snap2.get(ns);
  if (snap1.equals(snap2)) {
    break;
  } else if (attempts >= 10) {
    throw new Exception("Failed to get stable reading of metaspace performance counters after " + attempts + " tries");
  }
}```

(I do not know if it is correct or even compiles)

Reply:
Since this seems quite an "eye of the beholder" situation, and I do not really care either way, I'll adopt your suggestion.